### ISSUE TITLE
imap: Questions regarding ID extension

### ISSUE THREAD

**Issue by duesee** [modern-email/defects#12]
RFC 2971 uses ...

```abnf
id_params_list ::= "(" #(string SPACE nstring) ")" / nil
    ;; list of field value pairs
```

... to define the parameter list.

I believe that `::=` should be `=` (according to the current ABNF definition) and that `#` means "separated by whitespace"¹ (according to the examples).

Question: Does `#` mean `1*` ("at least one") or `*` ("zero or many")?

¹ I saw it described in other contexts as "separated by comma".

**Comment by cketti** [comment]
[RFC 2971](https://datatracker.ietf.org/doc/html/rfc2971) contains all the necessary references. 

> [4](https://datatracker.ietf.org/doc/html/rfc2971#section-4). Formal Syntax
>
>   This  syntax is intended to augment the grammar specified in [[IMAP4rev1](https://datatracker.ietf.org/doc/html/rfc2971#ref-IMAP4rev1)] in order to provide for the ID command. This specification uses the augmented Backus-Naur Form (BNF) notation as used in [[IMAP4rev1](https://datatracker.ietf.org/doc/html/rfc2971#ref-IMAP4rev1)].

The referenced IMAP4rev1 spec ([RFC 2060](https://datatracker.ietf.org/doc/html/rfc2060)) contains:

> [9](https://datatracker.ietf.org/doc/html/rfc2060#section-9). Formal Syntax
>
> The following syntax specification uses the augmented Backus-Naur Form (BNF) notation as specified in [[RFC-822](https://datatracker.ietf.org/doc/html/rfc822)] with one exception; the delimiter used with the "#" construct is a single space (SPACE) and not one or more commas.

The referenced [RFC 822](https://datatracker.ietf.org/doc/html/rfc822) has this to say:

>[2.7](https://datatracker.ietf.org/doc/html/rfc822#section-2.7).  #RULE:  LISTS
>     
> A construct "#" is defined, similar to "*", as follows:
>     
> &lt;l>#&lt;m>element
>     
> indicating at least &lt;l> and at most &lt;m> elements, each separated by one or more commas (","). This makes the usual form of lists very easy; a rule such as '(element *("," element))' can be shown as "1#element". Wherever this construct is used, null elements are allowed, but do not contribute to the count of elements present. That is, "(element),,(element)" is permitted, but counts as only two elements. Therefore, where at least one element is required, at least one non-null element must be present. Default values are 0 and infinity so that "#(element)" allows any number, including zero; "1#element" requires at least one; and "1#2element" allows one or two.

So the original

```
id_params_list ::= "(" #(string SPACE nstring) ")" / nil
```

could be written in modern notation as

```
field-value    = string SP nstring
id-params-list = "(" [field-value *(SP field-value)] ")" / nil
```

or

```
id-params-list = "(" [string SP nstring *(SP string SP nstring)] ")" / nil
```

if you prefer to have it all in one rule.

**Comment by duesee** [comment]
Thanks! I should have followed the RFC tracks more carefully.

**Comment by duesee** [comment]
Okay, that was fast ...

```imap
openssl s_client -crlf -quiet -connect imap.gmx.de:993
* OK [CAPABILITY IMAP4rev1 CHILDREN ENABLE ID IDLE LIST-EXTENDED LIST-STATUS LITERAL- MOVE NAMESPACE SASL-IR SORT SPECIAL-USE THREAD=ORDEREDSUBJECT UIDPLUS UNSELECT WITHIN AUTH=LOGIN AUTH=PLAIN] IMAP server ready H migmx112 27.4 IMAP-1McqiY-1rkrJv1mBW-00ZTj0
A ID nil
* ID ("name" "nemesis" "version" "27.4")
A OK ID completed
B ID ()
B BAD expected string (quoted / literal) instead of ")"
```

**Comment by duesee** [comment]
Dovecot and Gmail are fine with `A ID ()`.

**Comment by cketti** [comment]
Most people implementing the ID extension, probably don't spend much time thinking about the empty list case. `nil` has the advantage that it's used in the example for both the command and the response. So implementers most likely accounted for this case.

As a server I would expect the client to at least send a "name" value. Otherwise why send the ID command? Email clients typically don't display the server response to the user.

**Comment by duesee** [comment]
Yeah. The reason I'm asking is because I'm not sure how to encode the ID command in `imap-types`. (Currently working on https://github.com/duesee/imap-codec/pull/392.)

`Option<Vec<(key, value)>>` would be most precise. But given that GMX and Exchange choke on `A ID ()`, maybe it makes more sense to always encode an empty list as `nil`. But in this case we don't need the `Option<>`.

The RFC explicitly mentions proxies: When a proxy removes a particular list item (from a list with only one element), it could maintain the intent "client wanted to send a list" by forwarding `()`. Forwarding `nil` instead would mean "client didn't send a list".

**Comment by duesee** [comment]
I should explain a bit: `imap-types` want to be a "type-safe IMAP standard" that strictly encodes the IMAP syntax. Also as a reference.

Thus, I feel like allowing `()` (but still always sending `nil` in a default "quirk mode").

**Comment by duesee** [comment]
@chibenwa, I think you came across something similar :-)

**Comment by chibenwa** [comment]
Correct.

It was Office 365 failing if the server answer contains `()` but suceeded when it contains `NIL` instead.

The not fun fact was that it was completly crashing the connection...
